apps/noodle/librapi2/branches/0.9.3-3guada1/tools/man/synce-registry.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.33.
.TH SYNCE-REGISTRY "1" "February 2004" "synce-registry parent-key key-name value-name" "User Commands"
.SH NAME
synce-registry \- manipulate the registry on the device
.SH DESCRIPTION
synce-registry parent-key key-name value-name

.pp
With \fBsynce-registry\fR registry entries on the device could be queried.

.TP
parent-key
Could be one of
.IP 
HKCR or HKEY_CLASSES_ROOT
.IP
HKCU or HKEY_CURRENT_USER
.IP
HKLM or HKEY_LOCAL_MACHINE
.IP
HKU of HKEY_USERS

.TP
key-name
has to be an existing registry key

.TP
value-name
has to be an existing name of an registry value

.SH EXAMPLES

.LP
Query the current partner of the device
.IP
synce-registry HKLM "SOFTWARE\\Microsoft\\Windows CE Services\\Partners" PCur

.SH "AUTHOR"
.PP
This manual page was written by David Eriksson <twogood@users.sourceforge.net>
and updated by Volker Christian <voc@users.sourceforge.net>.
.SH "SEE ALSO"
synce(1) pcp(1) pls(1) prm(1) pmkdir(1) prmdir(1)


Changelog
BETA3 (n/a)
===========

PERSONALIZACIÓN Y CONFIGURACIÓN DEL SISTEMA (BRANDING)
------------------------------------------------------

- Cerrada lista prácticamente definitiva de aplicaciones
- Incluidas nuevas aplicaciones:
  - lsr
  - mozilla-acroread
  - dasher
  - gFTP (BTS 1087)
  - xchat (BTS 1086)
  - bluefish (BTS 1085)
  - d4x (BTS 1084)
  - soundtracker (BTS 1063)
  - gnome-games
  - firestarter
  - streamtuner
  - aMSN (v. 0.96)
  - gdebi
  - audacity (BTS 1059)
- Incluidas aplicaciones de consola (del spec): mc, ncftp, minicom, vim, ssh,
wget, rsync, dillo, sqlite, nmap, gsm-utils, brltty, joe
- Mejoras en el sistema de suplementos: comprobación de si el suplemento se
instala correctamente, comprobación de dependencias,...
- Adaptación de "contenido para Windows" de la v3 a la v4: pendiente de
terminar de depurar (BTS 1044)
- Solucionado problema de acceso a unidades SAMBA de la red: no se refrescaban
correctamente las unidades de red a través de nautilus
- Incluidos ficheros de ejemplo, adaptando los de la v3
- Pruebas de hibernación y gestión de energía en equipos portátiles correctas
- Añadido asistente para instalación de codecs de audio/vídeo
- Homogeneizadas entradas del menú: descripción (nombre aplicación)
- Implementado e incluido asistente para conexión a Internet
- Incluido "watermain"
- Traducido "gxine" (al final, descartada de la distribución base)
- Incluida imagen de fondo de Grub
- Personalizada página de prueba de impresión
- Configuración adecuada de "rhythmbox": inclusión de algunas radios
nacionales (tal y como se indica en el "spec")
- Configuración adecuada de "liferea": inclusión de algunos periódicos
nacionales (tal y como se indica en el "spec")
- Las aplicaciones en "Herramientas de Sistema" pasan a "Sistema ->
Administración"
- Configuración adecuada de "amule": navegador y reproductor de vídeo por
defecto (BTS 1017)


INSTALADOR LIVE (LIVE-INSTALLER)
--------------------------------

- Añadido visualizador de particiones de disco en selección de puntos de
montaje
- Mejorado (más usable y sencillo) el resumen final del instalador (BTS 955)
- Corregido problema de instalación de paquetes de soporte de inglés al final
de la instalación (ya no los instala)
- Corregido error de instalación cuando existen múltiples discos duros (BTS
965 y 1126)
- Mejorado sistema de redimensionado: permitía disminuir el tamaño por debajo
del mínimo necesario para la instalación
- Mejorado opción de utilizar espacio libre: permitía usar espacio sin
particionar demasiado pequeño
- Añadido soporte para instalación del perfil de accesibilidad con el que se
arranca en live al usuario que se crea en el sistema instalado
- Mejoradas descripciones de opciones de particionado
- Corregidos errores de unidades montadas tras usar "DiskPreview"
- Establecidos valores mínimos de instalación: 3 GB (/), 500 MB (/home), 64 MB
(swap)
- Análisis de soporte de accesibilidad de "gparted": habría que hacer un "fork"
- Personalizado botón de instalar incluyendo nuevo logo

INSTALADOR NO LIVE (DI-INSTALLER)
---------------------------------

- Corregidos algunos problemas leves de traducción

ACCESIBILIDAD
-------------

- Análisis y repaso de perfiles de accesibilidad
- Validación de funcionamiento de los perfiles de accesibilidad

NOODLE
------

- Implementado e incluido actor de Hermes para gestión de PDAs
- Finalizada integración en Guadalinex: paquetes guadalinex-noodle
- Análisis de la tecnología "opensync" para sincronización a través de
"syncml" (eg: symbian); probadas diferentes versiones (0.18 (dapper), 0.19,
020svn): descartado por considerarse poco estable aún.
- Corregidos algunos errores.

SISTEMA DE GENERACIÓN (GENERATION SYSTEM)
-----------------------------------------

- Actualizados paquetes relevantes de "-updates", "-security" y "-backports"
- Mejorada la estructura de metapaquetes de Guadalinex para hacer más flexible
la desinstalación de paquetes sin afectar a las configuraciones del resto

E-ADMIN
-------

- Incluido sistema de instalación de certificado raíz de la FNMT:
  - Cuando se detecta un certificado de usuario y se va a instalar en Firefox,
  se pide permiso al usuario para la instalación del certificado raíz,
  pidiendo que acepte la licencia.
  - Hecho lo anterior, se instala el certificado raíz en el sistema
- Modificado sistema de creación de usuarios para añadir a los nuevos usuarios
al grupo "scard"
- Añadidas librerías "jss" para acceso a diferentes servicios Java de la JA
- Análisis de firma en Evolution con DNIe: descartado (no es posible firmar
con el DNIe)


BETA2 (12/01/07)
================

PERSONALICACIÓN Y CONFIGURACIÓN DEL SISTEMA (BRANDING)
------------------------------------------------------

- Incluido "ndiswrapper" para soporte de algunas tarjetas wifi
- Añadidos iconos en el escritorio: "carpeta personal" y "equipo".
- Repaso e inclusión en Guadalinex de todos los paquetes interesantes que
incluye Ubuntu Edgy y que no estaban en Guadalinex: evolution-exchange,...
- Añadidos repositorios oficiales de Guadalinex v4
- Implementado interfaz de usuario para suplementos de manera que se pueda
utilizar sin hermes
- Compartición del directorio "Compartido" para permitir a los usuarios
compartir ficheros a través de SAMBA
- Regenerados "grub" y "grub-installer" para configurar adecuadamente el
nombre de la distribución ("Ubuntu" -> "Guadalinex")
- Comprobación de detección de entradas de GRUB para otros sistemas
operativos tras instalación
- Implementación de nueva funcionalidad para gnome-system-tools: soporte de
perfiles de accesibilidad para los usuarios que se crean
- Establecido número de escritorios virtuales a 4
- Añadido "alacarte" para edición de menús
- Añadido "camorama" para uso de webcams
- Añadido "lemurae" para acceso al diccionario de la RAE
- Traducido mensaje inicial al abrir una consola
- Personalizado el icono de salida de sesión
- Deshabilitado servidor SSH por defecto

apps/noodle/librapi2/trunk/tools/synce-registry.c
/* $Id: synce-registry.c 2348 2006-04-05 14:58:44Z twogood $ */
#include <rapi.h>
#include <synce_log.h>
#include <pcommon.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>

#define ACTION_READVAL 0
#define ACTION_WRITEVAL 1
#define ACTION_DELETEVAL 2
#define ACTION_LISTKEY 3
#define ACTION_NEWKEY 4
#define ACTION_DELETEKEY 5

char* devpath = NULL;
int action = ACTION_READVAL;
DWORD valType = REG_SZ;
const char *prog_name;

#define STR_EQUAL(a,b)  (0 == strcasecmp(a,b))

static void show_usage(const char* name)
{
	fprintf(stderr,
			"Syntax:\n"
			"\n"
			"\t%s [-d LEVEL] [-p DEVPATH] [-t TYPE ] [-h]\n"
                        "\t[-r] PARENTKEY KEY VALUE\t\tRead value\n"
                        "\t  -w PARENTKEY KEY VALUE NEWVALUE\tWrite value\n"
                        "\t  -l PARENTKEY KEY VALUE\t\tList key\n"
                        "\t  -n PARENTKEY KEY\t\t\tNew key\n"
                        "\t  -x PARENTKEY KEY VALUE\t\tDelete value (not supported)\n"
                        "\t  -X PARENTKEY KEY\t\t\tDelete key (not supported)\n"
			"\n"
			"\t-d LEVEL     Set debug log level\n"
			"\t                 0 - No logging (default)\n"
			"\t                 1 - Errors only\n"
			"\t                 2 - Errors and warnings\n"
			"\t                 3 - Everything\n"
			"\t-h           Show this help message\n"
                        "\t-p DEVPATH   Device path\n"
                        "\t-t TYPE      New type for writes:\n"
                        "\t\tsz         String (default)\n"
                        "\t\tdword      Double-word\n"
                        "\t\tdword_bige Big-endian double-word\n"
                        "\t\texpand_sz  String with path expansion\n"
                        "\t\tmulti_sz   Multiple strings (not supported)\n"
                        "\t\tbinary     Binary data (not supported)\n"
			"\tPARENTKEY    Registry parent (HKLM, etc.)\n"
			"\tKEY          Registry key\n"
			"\tVALUE        Registry value within key\n"
			"\tNEWVALUE     New value for writes\n"
                        "",
			name
                );
}


static void dump(const char *desc, void* data, size_t len)/*{{{*/
{
	uint8_t* buf = (uint8_t*)data;
	size_t i, j;
	char hex[8 * 3 + 1];
	char chr[8 + 1];

	printf("%s (%d bytes):\n", desc, len);
	for (i = 0; i < len + 7; i += 8) {
		for (j = 0; j < 8; j++) 
			if (j + i >= len) {
				hex[3*j+0] = ' ';
				hex[3*j+1] = ' ';
				hex[3*j+2] = ' ';
				chr[j] = ' ';
			} else {
				uint8_t c = buf[j + i];
				const char *hexchr = "0123456789abcdef";
				hex[3*j+0] = hexchr[(c >> 4) & 0xf];
				hex[3*j+1] = hexchr[c & 0xf];
				hex[3*j+2] = ' ';
				if (c > ' ' && c <= '~')
					chr[j] = c;
				else
					chr[j] = '.';
			}
		hex[8*3] = '\0';
		chr[8] = '\0';
		printf("  %04x: %s %s\n", i, hex, chr);
	}
}/*}}}*/

static bool handle_parameters(int argc, char** argv, char** parent_str, char** key_name, char** value_name, char **new_value)
{
	int c;
	int log_level = SYNCE_LOG_LEVEL_LOWEST;
	while ((c = getopt(argc, argv, "d:hp:t:rwxlnX")) != -1)
	{
		switch (c)
		{
			case 'd':
				log_level = atoi(optarg);
				break;

                        case 'p':
                                devpath = optarg;
                                break;
			
                        case 'r':
                                action = ACTION_READVAL;
                                break;

                        case 'w':
                                action = ACTION_WRITEVAL;
                                break;

                        case 'x':
                                action = ACTION_DELETEVAL;
                                break;

                        case 'l':
                                action = ACTION_LISTKEY;
                                break;

                        case 'n':
                                action = ACTION_NEWKEY;
                                break;

                        case 'X':
                                action = ACTION_DELETEKEY;
                                break;

                        case 't':
                                if (strcasecmp(optarg,"binary") == 0)
                                {
                                  valType = REG_BINARY;
                                }
                                else if (strcasecmp(optarg,"dword")==0 || strcasecmp(optarg,"dword_little_endian") == 0)
                                {
                                  valType = REG_DWORD;
                                }
                                else if (strcasecmp(optarg,"dword_bige") == 0)
                                {
                                  valType = REG_DWORD_BIG_ENDIAN;
                                }
                                else if (strcasecmp(optarg,"expand_sz") == 0)
                                {
                                  valType = REG_EXPAND_SZ;
                                }
                                else if (strcasecmp(optarg,"multi_sz") == 0)
                                {
                                  valType = REG_MULTI_SZ;
                                }
                                else if (strcasecmp(optarg,"sz") == 0)
                                {
                                  valType = REG_SZ;
                                }
                                else
                                {
                                  fprintf(stderr,"%s: Unrecognized value type '%s' for argument -t\n",argv[0],optarg);
                                  return false;
                                }
                                break;

			case 'h':
			default:
				show_usage(argv[0]);
				return false;
		}
	}
				
	synce_log_set_level(log_level);

        argc -= optind;
        switch(action)
        {
          case ACTION_READVAL:
          case ACTION_DELETEVAL:
            if (argc != 3)
            {
              fprintf(stderr,"Wrong number of parameters\n");
              show_usage(argv[0]);
              return false;
            }
            break;
          case ACTION_WRITEVAL:
            if (argc != 4)
            {
              fprintf(stderr,"Wrong number of parameters\n");
              show_usage(argv[0]);
              return false;
            }
            break;
          case ACTION_DELETEKEY:
          case ACTION_LISTKEY:
            if (argc != 2)
            {
              fprintf(stderr,"Wrong number of parameters\n");
              show_usage(argv[0]);
              return false;
            }
            break;
        }


        switch(argc)
        {
          case 4:
            if ((*new_value = strdup(argv[optind+3])) == NULL)
              return false;
            /* fall through */
          case 3:
            if ((*value_name = strdup(argv[optind+2])) == NULL)
              return false;
            /* fall through */
          case 2:
            if ((*key_name = strdup(argv[optind+1])) == NULL)
              return false;
            /* fall through */
          case 1:
            if ((*parent_str = strdup(argv[optind])) == NULL)
              return false;
        }

	return true;
}

int delete_key(HKEY parent, LPCWSTR key_name_wide)
{
#if 0
  int error;
  
  if (ERROR_SUCCESS != (error=CeRegDeleteKey(parent, key_name_wide)))
  {
    fprintf(stderr,"Failed to delete key '%s': %s\n",
            wstr_to_ascii(key_name),
            synce_strerror(error));
    return 1;
  }
  return 0;
#else
  fprintf(stderr,"CeRegDeleteKey not yet implemented\n");
  return 1;
#endif      
}

int new_key(HKEY parent, LPCWSTR key_name_wide)
{
  int error;
  HKEY new_key;
  DWORD new_disposition;
  
  if (ERROR_SUCCESS != (error=CeRegCreateKeyEx(parent, key_name_wide, 0,0,0,0,NULL,&new_key,&new_disposition)))
  {
    fprintf(stderr,"Failed to create key '%s': %s\n",
            wstr_to_ascii(key_name_wide),
            synce_strerror(error));
    return 1;
  }
  if (ERROR_SUCCESS != (error=CeRegCloseKey(new_key)))
  {
    fprintf(stderr,"Failed to close key '%s': %s\n",
            wstr_to_ascii(key_name_wide),
            synce_strerror(error));
    return 1;
  }

  if (new_disposition != REG_CREATED_NEW_KEY)
  {
    fprintf(stderr,"Key '%s' already exists\n",
            wstr_to_ascii(key_name_wide));
    return 1;
  }
  
  return 0;
}

int list_key(HKEY key)
{
  int result;
  int i;

  for(i = 0; ; i++)
  {
    WCHAR wide_name[MAX_PATH];
    DWORD name_size = sizeof(wide_name);
    
    result = CeRegEnumKeyEx(key, i, wide_name, &name_size, NULL, NULL,
                            NULL, NULL);
    if (ERROR_NO_MORE_ITEMS == result)
      break;
    else if (ERROR_SUCCESS != result)
    {
      fprintf(stderr,"Failed to list subkey #%d: %s\n",
              i,
              synce_strerror(result));
      return 1;
    }
    printf("K %s\n",wstr_to_ascii(wide_name));
  }

  for(i = 0; ; i++)
  {
    WCHAR wide_name[MAX_PATH];
    DWORD name_size = sizeof(wide_name);
    DWORD type;
    
    result = CeRegEnumValue(key, i, wide_name, &name_size, NULL, &type,
                            NULL, NULL);
    if (ERROR_NO_MORE_ITEMS == result)
      break;
    else if (ERROR_SUCCESS != result)
    {
      fprintf(stderr,"Failed to list subkey #%d: %s\n",
              i,
              synce_strerror(result));
      return 1;
    }
    /* TODO: We already have the value type, might as well display it. */
    /* TODO: We have the value itself, too, so maybe just display that. */
    /* TODO: It's probably pretty easy to make this work recursively, while we're at it */
    /* TODO: Which would be a pretty good registry dump tool. */
    printf("V %s\n",wstr_to_ascii(wide_name));
  }
  return 0;
}

int delete_val(HKEY key, LPCWSTR value_name)
{
#if 0
  int error;
  
  if (ERROR_SUCCESS != (error=CeRegDeleteKey(key, value_name)))
  {
    fprintf(stderr,"Failed to delete value '%s': %s\n",
            wstr_to_ascii(value_name),
            synce_strerror(error));
    return 1;
  }
  return 0;
#else
  fprintf(stderr,"CeRegDeleteValue not yet implemented\n");
  return 1;
#endif      

}

int read_val(char *parent_str, char *key_str, HKEY key, LPCWSTR value_name_wide)
{
  int error;
  DWORD value_size = 0;
  uint8_t* value = NULL;
  DWORD value_type;

  if (ERROR_SUCCESS != (error=CeRegQueryValueEx(key, value_name_wide, NULL, NULL, NULL, &value_size)))
  {
    fprintf(stderr, "Failed to get size of value '%s\\%s\\%s': %s\n", 
        parent_str, key_str, wstr_to_ascii(value_name_wide),
        synce_strerror(error));
    return 1;
  }

  value = calloc(1, value_size);
  if (!value)
  {
    fprintf(stderr, "Failed to allocate %i bytes", value_size);
    return 1;
  }
  
  if (ERROR_SUCCESS != (error=CeRegQueryValueEx(key, value_name_wide, NULL, &value_type, value, &value_size)))
  {
    fprintf(stderr, "Failed to get value '%s\\%s\\%s': %s\n", 
        parent_str, key_str, wstr_to_ascii(value_name_wide),
        synce_strerror(error));
    return 1;
  }

  printf("[%s\\%s]\n\"%s\"=", parent_str, key_str, wstr_to_ascii(value_name_wide));

  switch (value_type)
  {
    case REG_SZ:
      printf("\"%s\"\n", wstr_to_ascii((LPCWSTR)value));
      break;
    
    case REG_DWORD:
      printf("dword=%08x\n", *(DWORD*)value);
      break;

    default:
      dump("Value", value, value_size);
      break;
  }

  return 0;
}


int write_val(HKEY key, LPCWSTR value_name, char * new_value)
{
  int error;
  void *valBuf;
  int valSize;
  DWORD dwordVal;
  
  switch(valType)
  {
    case REG_SZ:
    case REG_EXPAND_SZ:
      valBuf = wstr_from_current(new_value);
      valSize = (wstrlen(valBuf)+1) * sizeof(WCHAR);
      break;
    case REG_DWORD:
      dwordVal = strtol(new_value,NULL,0);
      valBuf = &dwordVal;
      valSize = sizeof(DWORD);
      break;
    case REG_DWORD_BIG_ENDIAN:
      fprintf(stderr,"REG_DWORD_BIG_ENDIAN not yet supported.\n");
      return 1;
    case REG_MULTI_SZ:
      fprintf(stderr,"REG_MULTI_SZ not yet supported.\n");
      return 1;
    default:
      fprintf(stderr,"Unrecognized value type somehow!\n");
      return 1;
  }
  if (ERROR_SUCCESS != (error=CeRegSetValueEx(key,value_name,0,valType,valBuf,valSize)))
  {
    fprintf(stderr, "%s: Unable to set value of '%s' to REG_SZ '%s': %s\n", 
        prog_name,wstr_to_ascii(value_name),new_value,
        synce_strerror(error));
    return 1;
  }
  return 0;
}

int main(int argc, char** argv)
{
  int result = 1;
  RapiConnection* connection = NULL;
  char* parent_str  = NULL;
  char* key_name          = NULL;
  char* value_name  = NULL;
  char* new_value   = NULL;
  HKEY parent;
  HKEY key = 0;
  WCHAR* value_name_wide = NULL;
  WCHAR* key_name_wide = NULL;
  HRESULT hr;

  prog_name = argv[0];
  
  if (!handle_parameters(argc,argv,&parent_str,&key_name,&value_name,&new_value))
    goto exit;
  
  /* handle abbreviations */
  if (STR_EQUAL(parent_str, "HKCR"))
    parent_str = "HKEY_CLASSES_ROOT";
  else if (STR_EQUAL(parent_str, "HKCU"))
    parent_str = "HKEY_CURRENT_USER";
  else if (STR_EQUAL(parent_str, "HKLM"))
    parent_str = "HKEY_LOCAL_MACHINE";
  else if (STR_EQUAL(parent_str, "HKU"))
    parent_str = "HKEY_USERS";

  if (STR_EQUAL(parent_str, "HKEY_CLASSES_ROOT"))
    parent = HKEY_CLASSES_ROOT;
  else if (STR_EQUAL(parent_str, "HKEY_CURRENT_USER"))
    parent = HKEY_CURRENT_USER;
  else if (STR_EQUAL(parent_str, "HKEY_LOCAL_MACHINE"))
    parent = HKEY_LOCAL_MACHINE;
  else if (STR_EQUAL(parent_str, "HKEY_USERS"))
    parent = HKEY_USERS;
  else
  {
    fprintf(stderr, "Invalid parent key\n");
    goto exit;
  }

  if ((connection = rapi_connection_from_path(devpath)) == NULL)
  {
    fprintf(stderr, "%s: Could not find configuration at path '%s'\n", 
        argv[0],
        devpath?devpath:"(Default)");
    goto exit;
  }
  rapi_connection_select(connection);
  if (S_OK != (hr = CeRapiInit()))
  {
    fprintf(stderr, "%s: Unable to initialize RAPI: %s\n", 
        argv[0],
        synce_strerror(hr));
    goto exit;
  }

  convert_to_backward_slashes(key_name);
  key_name_wide = wstr_from_current(key_name);

  if (action == ACTION_DELETEKEY)
  {
    result = delete_key(parent, key_name_wide);
    goto exit;
  }
  else if (action == ACTION_NEWKEY)
  {
    result = new_key(parent, key_name_wide);
    goto exit;
  }

  if (!rapi_reg_open_key(parent, key_name, &key))
  {
    fprintf(stderr, "Failed to open key: '%s\\%s'\n", parent_str, key_name);
    goto exit;
  }

  if (action == ACTION_LISTKEY)
  {
    result = list_key(key);
    goto exit;
  }

  value_name_wide = wstr_from_current(value_name);

  if (action == ACTION_DELETEVAL)
  {
    result = delete_val(key, value_name_wide);
  }
  else if (action == ACTION_READVAL)
  {
    result = read_val(parent_str, key_name, key, value_name_wide);
  }
  else if (action == ACTION_WRITEVAL)
  {
    result = write_val(key, value_name_wide, new_value);
  }
  
exit:
  if (key)
    CeRegCloseKey(key);
  CeRapiUninit();
  if (key_name)
    free(key_name);
  if (key_name_wide)
    wstr_free_string(key_name_wide);
  if (value_name_wide)
    wstr_free_string(value_name_wide);
  return result;
}

